File: FruitWholesale.Mobile/lib/features/shop_return/shop_return_service.dart
import '../../core/api/api_client.dart';
import '../../core/models/paginated_list.dart';
import 'shop_return_models.dart';

class ShopReturnService {
  final ApiClient _api;

  ShopReturnService(this._api);

  Future<PaginatedList<ShopReturnListItem>> getPaged({
    int pageNumber = 1,
    int pageSize = 20,
    String? searchTerm,
    int? shopId,
    String? fromDate,
    String? toDate,
  }) async {
    final json = await _api.get('/shopreturn', query: {
      'pageNumber': pageNumber,
      'pageSize': pageSize,
      if (searchTerm != null && searchTerm.isNotEmpty) 'searchTerm': searchTerm,
      if (shopId != null) 'shopId': shopId,
      if (fromDate != null) 'fromDate': fromDate,
      if (toDate != null) 'toDate': toDate,
    });
    return PaginatedList.fromJson(json as Map<String, dynamic>, ShopReturnListItem.fromJson);
  }

  Future<ShopReturnDetail> getById(int id) async {
    final json = await _api.get('/shopreturn/$id');
    return ShopReturnDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<String> getNextReferenceNo() => _api.getText('/shopreturn/next-reference-no');

  Future<ShopReturnDetail> create(ShopReturnDetail shopReturn) async {
    final json = await _api.post('/shopreturn', body: shopReturn.toSaveJson());
    return ShopReturnDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<ShopReturnDetail> update(int id, ShopReturnDetail shopReturn) async {
    final json = await _api.put('/shopreturn/$id', body: shopReturn.toSaveJson());
    return ShopReturnDetail.fromJson(json as Map<String, dynamic>);
  }
}

File: FruitWholesale.Client/src/app/features/shop-return/shop-return.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { PaginatedList, PaginationRequest } from '../../core/models/common.model';
import { SaveShopReturn, ShopReturn, ShopReturnListItem } from '../../core/models/transactions.model';
import { toHttpParams } from '../../core/utils/http-params.util';

@Injectable({ providedIn: 'root' })
export class ShopReturnService {
  private readonly baseUrl = `${environment.apiUrl}/shopreturn`;

  constructor(private readonly http: HttpClient) {}

  getPaged(request: PaginationRequest, shopId?: number | null, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<ShopReturnListItem>>(this.baseUrl, {
      params: toHttpParams({ ...request, shopId, fromDate, toDate })
    });
  }

  getById(id: number): Observable<ShopReturn> {
    return this.http.get<ShopReturn>(`${this.baseUrl}/${id}`);
  }

  getNextReferenceNo(): Observable<string> {
    return this.http.get(`${this.baseUrl}/next-reference-no`, { responseType: 'text' });
  }

  create(dto: SaveShopReturn): Observable<ShopReturn> {
    return this.http.post<ShopReturn>(this.baseUrl, dto);
  }

  update(id: number, dto: SaveShopReturn): Observable<ShopReturn> {
    return this.http.put<ShopReturn>(`${this.baseUrl}/${id}`, dto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
